odl/test/util/numerics_test.py

```diff
@@ -482,5 +573,48 @@ def test_resize_array_raise():
         resize_array(small_arr, (3, 4), offset=(0, 1), pad_mode='periodic')
 
 
+def test_binning(binning_setup):
+    """Validate the ``binning`` function."""
+    bin_size, bin_reduction, _, arr, true_out_arr = binning_setup
+
+    # Error scenarios
+    bin_sizes = bin_size if isinstance(bin_size, tuple) else [bin_size] * 2
+    if 0 in bin_sizes:
+        # 0 not allowed
+        with pytest.raises(ValueError):
+            binning(arr, bin_size, bin_reduction)
+        return
+    if any(b > n for b, n in zip(bin_sizes, arr.shape)):
+        # Bin size larger than array size not allowed
+        with pytest.raises(ValueError):
+            binning(arr, bin_size, bin_reduction)
+        return
+    if any(n % b != 0 for b, n in zip(bin_sizes, arr.shape)):
+        # Division remainder of size/bin_size not allowed
+        with pytest.raises(ValueError):
+            binning(arr, bin_size, bin_reduction)
+        return
+
+    # Success scenarios
+    out_arr = binning(arr, bin_size, bin_reduction)
+    assert all_almost_equal(out_arr, true_out_arr)
+    assert out_arr.dtype == true_out_arr.dtype
+
+
+def test_binning_corner_cases():
+    # 0-dim
+    arr = np.array(1.0)
+    assert binning(arr, 2) == 1.0  # ok since there are no axes
+    with pytest.raises(ValueError):
+        binning(arr, [2])  # not ok, too many binning values
```

Comment:
I fail to see the difference between the two cases: why is the first one ok, but the second one is not?

Reply:
It's largely a consequence of the implementation. `binning(arr, scalar)` is equivalent to `binning(arr, [scalar] * arr.ndim)`, which is `binning(arr, [])` in the first case. It results in something like `np.sum(arr.reshape(()), axis=())`, which seems to work fine.
The other one is clearly a violation because `bin_sizes` is longer than `ndim`.

Reply:
I'd argue that this is consistent behavior: "reduce over all axes" just means "don't reduce" if there are no axes.